File: intro_to_NA/file/diffusion1d.py
## Import Modules
import numpy as np
from scipy.stats import hmean
import matplotlib.pyplot as plt

## Input Parmeters
L = 2*np.pi # Length of Reservoir[m]
N = 100 # Number of Control Volume[-]
k = 1*np.ones(N) # Permiability[m^2]
phi = np.ones(N) # Porosity[m^2]
c = np.ones(N) # Compressibility[Pa^-1]
mu= 1 # Viscosity of Fluid[Pa^-1]
dx = L/N # Size of Control Volume[m]
x = np.zeros(N) # x coordinate[m]
x[0] = dx/2
for i in range(1, N):
  x[i] = x[i-1] + dx

## Parameters for Simulation
tmax = 20 # Time to stop simlation [s]
dt   = 0.005   # dt [s]
nout = 500  # output result ever nout step[s]
# Variable to decide Boundary condition
#  1-> Neumann, 0-> Dirichlet
BC_right = 0 # Boundary at right(x = L) 
BC_left  = 0 # Boundary at Left (x = 0)
Pb_right = 0 # Pressure Value at x = L
Pb_left  = 0 # Pressure Value at x = 0

## Check dt meets conditions to converge.
#-# Write Your Code Here #-#

## Initial Conditions
# P_init = np.ones(N)      # Initial Pressure
P_init = np.sin(x)

## Simulation
P_old  = np.copy(P_init) # Pressure at n-th step
P_new  = np.copy(P_init) # Pressure at n+1-th step
t = 0
n = 0
plt.plot(x, P_new, label='t={0:05.2f}'.format(t)) # Plot Initial Condition
while True:
  for i in range(1, N-1): # for P[1] ~ P[N-2]
    alpha = dt / (phi[i]*c[i])
    lam_w = hmean([k[i-1], k[i]])/mu
    lam_e = hmean([k[i+1], k[i]])/mu
    A = #-# Write Your Code Here #-#
    B = #-# Write Your Code Here #-#
    C = #-# Write Your Code Here #-#
    P_new[i] = A*P_old[i-1] + B*P_old[i] + C*P_old[i+1]
  
  # P[0] CV#1
  if BC_left == 1: # Neumann Condition
    alpha = dt / (phi[i]*c[i])
    lam_w = hmean([k[i-1], k[i]])/mu
    lam_e = hmean([k[i+1], k[i]])/mu
    P_new[0] = #-# Write Your Code Here #-#
  else: # Dirichlet Condition
    alpha = dt / (phi[i]*c[i])
    lam_w = hmean([k[i-1], k[i]])/mu
    lam_e = hmean([k[i+1], k[i]])/mu
    P_new[0] = #-# Write Your Code Here #-#

  # P[N-1] #CV#N
  if BC_left == 1: # Neumann Condition
    P_new[N-1] = #-# Write Your Code Here #-#
  else: # Dirichlet Condition 
    P_new[N-1] = #-# Write Your Code Here #-#

  # Update Values, time step and Add plot
  P_old = np.copy(P_new)
  t = t + dt
  n = n + 1

  if n%nout == 0:
    print('{0}th Time step {1:05.2f}'.format(n, t))
    plt.plot(x, P_new, label='t={0:05.2f}'.format(t))
  
  if t >= tmax:
    break
    
plt.plot(x, P_new, label='t={0:05.2f}'.format(t)) # Plot Final Distribution
plt.title('Pressure Diffusion 1D')
plt.ylabel('Pressure [Pa]')
plt.xlim(0, L)
plt.ylim(-1,1)
plt.xlabel('x[m]')
plt.legend()
plt.show()

File: intro_to_NA/file/Diffusion1D.m
%% Input Parmeters
L = 2*pi; % Length of Reservoir(m)
N = 100; % Number of Control Volume(-)
k = 0.2*ones(1,N); % Permiability(m^2)
phi = ones(1,N); % Porosity(m^2)
c = ones(1,N); % Compressibility(Pa^-1)
mu = 1; % Viscosity of Fluid(Pa^-1)
dx = L/N; % Size of Control Volume(m)
x = zeros(1,N); % x coordinate(m)
x(1) = dx/2;
for i = 2:N
  x(i) = x(i-1) + dx;
end

%% Parameters for Simulation
tmax = 20; % Time to stop simlation (s)
dt   = 0.005;   % dt (s)
nout = 500;  % output result ever nout step(s)
% Variable to decide Boundary condition
% 1 -> Neumann, 0 -> Dirichlet
B_right  = 0;% Boundary at right (x = L)
B_left   = 0;% Boundary at left (x = 0)
Pb_right = 0;% Pressure Value at right (x = L) 
Pb_left  = 0;% Pressure Value at left (x = 0)

%% Initial Conditions
% P_init = ones(N)      % Initial Pressure
P_init = sin(x);

%% Simulation
P_old  = P_init; % Pressure at n-th step
P_new  = P_init; % Pressure at n+1-th step
t = 0;
n = 0;

ax1 = subplot(1,1,1);
plot(ax1, x, P_init, 'DisplayName',[num2str(t,'%05.2f'),'[s]'])
xlim(ax1, [0, L]);
ylim(ax1, [-1,1]);
title(ax1, 'Pressure Diffusion 1D');
hold(ax1, 'on');
while true
    for i = 2:N-1
        alpha = %-% Write Your Code Here %-%
        lam_w = %-% Write Your Code Here %-%
        lam_e = %-% Write Your Code Here %-%
        % A = %-% Write Your Code Here %-%
        % B = %-% Write Your Code Here %-%
        % C = %-% Write Your Code Here %-%
        P_new(i) = A*P_old(i+1) + B*P_old(i) + C*P_old(i-1);
    end
    
    % P(1)
    if B_left == 1 % Neumann Condition
        P_new(1) = %-% Write Your Code Here %-%

    else % Dirichlet Condition
        P_new(1) = %-% Write Your Code Here %-%
    end

    % P(N);
    if B_right == 1 % Neumann Condition;
        P_new(N) = %-% Write Your Code Here %-%

    else % Dirichlet Condition ;
        P_new(N) =%-% Write Your Code Here %-%
    end

    
    % Update Values, time step and Loop counter
    P_old = P_new;
    t = t + dt;
    n = n + 1;
    if t >= tmax
      break
    end
    
    if mod(n,nout) == 0
        plot(ax1, x, P_new, 'DisplayName',[num2str(t,'%05.2f'),'[s]'])
        hold(ax1, 'on');
        disp([num2str(n, '%04d'),'th time step',num2str(t, '%05.2f')])
    end
end

legend(ax1)